### PR TITLE
Fix Hovercard when using shadow DOM

### LINKED ISSUE #2983: Menu flickering inside shadow dom
### Current behavior

The <Menu> component's submenus flicker when used inside shadow dom. Which makes it hard to select a submenu item.

https://github.com/ariakit/ariakit/assets/30146019/64c978af-15d5-49a0-a310-5d102bd84ac1

### Steps to reproduce the bug

1. Open sandbox: https://stackblitz.com/edit/ondrbq?file=app%2Fpage.tsx
2. Open the menu and mouse the cursor over the items with a submenu.

In the example above is the same as the default demo, expect that I've wrapped the menu example in `ShadowRoot.div` to render it inside shadow dom, and (optional) moved the styles inside the shadow dom, so the menu looks presentable.

This is not related to `react-shadow`, this can be reproduced also by manually creating a shadow dom and rendering the menu there.

### Expected behavior

The menu should work without flickering.

### Workaround

https://github.com/ariakit/ariakit/issues/2983#issuecomment-1775092916

### Possible solutions

https://github.com/ariakit/ariakit/issues/2983#issuecomment-1775092916

### PR DESCRIPTION
**Pull request by georgekaran**
Closes #2983

### CODE REVIEW

File: examples/hovercard-shadow-dom/index.tsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  const [count, setCount] = useState(0);
+  return (
+    <Fragment>
+      <span>Opened {count} times</span>
+      <ShadowRootHost>
+        <Ariakit.HovercardProvider
+          setOpen={(open) => {
+            if (open) setCount(count + 1);
+
+            return open;
```
Comment (diegohaz) [inline examples/hovercard-shadow-dom/index.tsx]: Returning is not necessary.
```suggestion
            if (open) setCount(count + 1);
```

File: examples/hovercard-shadow-dom/test-browser.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+const getHovercard = (page: Page) =>
+  page.getByRole("dialog", { name: "Ariakit" });
+
+test.beforeEach(async ({ page }) => {
+  await page.goto("/previews/hovercard-shadow-dom", {
+    waitUntil: "networkidle",
+  });
+});
+
+test.describe.configure({ retries: process.env.CI ? 2 : 1 });
```
Comment (diegohaz) [inline examples/hovercard-shadow-dom/test-browser.ts]: I suppose this is coming from another test, right? Did you notice any flakiness on CI? If not, I believe we can remove this.

File: examples/hovercard-shadow-dom/test-browser.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  await getAnchor(page).hover({ position: { y: 10, x: 10 } });
+  await expect(page.getByText("Opened 1 times")).toBeVisible();
+  await getAnchor(page).hover({ position: { y: 10, x: 20 } });
+  await expect(page.getByText("Opened 1 times")).toBeVisible();
+  await page.mouse.move(0, 0);
+  await expect(getHovercard(page)).not.toBeVisible();
+  await getAnchor(page).hover();
+  await expect(getHovercard(page)).toBeVisible();
+  await expect(page.getByText("Opened 2 times")).toBeVisible();
+});
```
Comment (diegohaz) [inline examples/hovercard-shadow-dom/test-browser.ts:28]: Have you run this test on the code without the fix to confirm we're testing the correct aspect (it's supposed to fail)?

Comment (georgekaran) [reply examples/hovercard-shadow-dom/test-browser.ts:28]: Yes sir, I wrote this test before applying my changes and it was failing as expected.

File: examples/hovercard-shadow-dom/test-browser.ts (hunk 2 above)
Comment (georgekaran) [reply examples/hovercard-shadow-dom/test-browser.ts]: Yeah, I copy this from the `hovercard example` but I did not experience any flakiness, so I'm gonna remove it for now.

File: examples/hovercard-shadow-dom/test-browser.ts (hunk 3 above)
Comment (diegohaz) [reply examples/hovercard-shadow-dom/test-browser.ts:28]: Awesome!

**Review by diegohaz — APPROVED** [review]

**Pull request merged**